File: web/src/state/uiActions.ts
// The ONE place UI actions from /api/ask change the app. Pure + tested. Unknown dish ids and unknown
// action types are ignored (actions carry ids only; numbers are always refetched from the engine).
import type { Deltas, Panel, UiAction } from "../contract";

export interface UiSlice {
  selectedId: string | null;
  panel: Panel;
  twinHighlight: string | null;
  shiftDeltas: Deltas;
  explainPair: [string, string] | null;
  highlightIds: string[];
}

export const initialUi: UiSlice = {
  selectedId: null,
  panel: "twins",
  twinHighlight: null,
  shiftDeltas: {},
  explainPair: null,
  highlightIds: [],
};

export function reduceUiAction(s: UiSlice, a: UiAction, known: Set<string>): UiSlice {
  const ok = (id: string | null | undefined): id is string => !!id && known.has(id);
  switch (a.type) {
    case "select_dish":
      return ok(a.dish_id) ? { ...s, selectedId: a.dish_id } : s;
    case "show_twins":
      return ok(a.dish_id)
        ? { ...s, selectedId: a.dish_id, twinHighlight: ok(a.highlight_dish_id) ? a.highlight_dish_id : null }
        : s;
    case "set_shift":
      return ok(a.dish_id) ? { ...s, selectedId: a.dish_id, shiftDeltas: { ...(a.deltas as Deltas) } } : s;
    case "explain_pair":
      return ok(a.a) && ok(a.b) ? { ...s, explainPair: [a.a, a.b] } : s;
    case "open_panel":
      return { ...s, panel: a.panel };
    case "highlight_dishes":
      return { ...s, highlightIds: a.dish_ids.filter((id) => known.has(id)) };
    default:
      return s; // forward compatible: ignore action types this build doesn't know
  }
}

export const reduceUiActions = (s: UiSlice, actions: UiAction[], known: Set<string>): UiSlice =>
  actions.reduce((acc, a) => reduceUiAction(acc, a, known), s);

File: contracts/src/tastespace_contracts/ui_actions.py
"""UI actions returned by /api/ask. The web app applies them with one reducer.

Rules (enforced server-side in backend/tastespace/grok): actions carry IDs only, never scores; every
referenced dish_id must exist AND must have appeared in that turn's tool_trace. The UI refetches all
numbers from the engine. Unknown action types must be ignored by the UI (forward compatible).
"""

from typing import Annotated, Literal, Union

from pydantic import BaseModel, ConfigDict, Field

from .taxonomy import DimId, Panel

ShiftDelta = Annotated[float, Field(ge=-2.0, le=2.0, description="sigma units")]


class _Action(BaseModel):
    model_config = ConfigDict(extra="forbid", json_schema_serialization_defaults_required=True)


class SelectDish(_Action):
    type: Literal["select_dish"] = "select_dish"
    dish_id: str


class ShowTwins(_Action):
    type: Literal["show_twins"] = "show_twins"
    dish_id: str
    highlight_dish_id: str | None = None


class SetShift(_Action):
    type: Literal["set_shift"] = "set_shift"
    dish_id: str
    deltas: dict[DimId, ShiftDelta]


class ExplainPair(_Action):
    type: Literal["explain_pair"] = "explain_pair"
    a: str
    b: str


class OpenPanel(_Action):
    type: Literal["open_panel"] = "open_panel"
    panel: Panel


class HighlightDishes(_Action):
    type: Literal["highlight_dishes"] = "highlight_dishes"
    dish_ids: list[str] = Field(min_length=1, max_length=10)


UiAction = Annotated[
    Union[SelectDish, ShowTwins, SetShift, ExplainPair, OpenPanel, HighlightDishes],
    Field(discriminator="type"),
]


def referenced_dish_ids(action: BaseModel) -> list[str]:
    """Every dish id an action points at (used by the grounding check)."""
    if isinstance(action, (SelectDish, SetShift)):
        return [action.dish_id]
    if isinstance(action, ShowTwins):
        return [action.dish_id] + ([action.highlight_dish_id] if action.highlight_dish_id else [])
    if isinstance(action, ExplainPair):
        return [action.a, action.b]
    if isinstance(action, HighlightDishes):
        return list(action.dish_ids)
    return []
